pkg/engine/combat/hit.go

```diff
@@ -27,22 +29,82 @@ func (mgr *Manager) performHit(hit *info.Hit) {
 	// * dots & element damage do not crit (unknown if also ByPureDamage?)
 	// * ByPureDamage = true means a "simplified" damage function (the break damage equation)
 
+	base := mgr.baseDamage(hit)*hit.HitRatio + hit.DamageValue
+	bonus := mgr.bonusDamage(hit)
+	crit := mgr.crit(hit)
+	total := mgr.totalDamage(hit, base, bonus, crit)
+	hpUpdate := mgr.shld.AbsorbDamage(hit.Defender.ID(), total)
+
+	mgr.attr.ModifyHPByAmount(hit.Defender.ID(), hit.Attacker.ID(), total, true)
+	mgr.attr.ModifyStance(hit.Defender.ID(), hit.Attacker.ID(), hit.StanceDamage*hit.HitRatio)
+	if mgr.target.IsCharacter(hit.Attacker.ID()) {
+		mgr.attr.ModifyEnergy(hit.Attacker.ID(), hit.EnergyGain*hit.HitRatio)
+	} else {
+		mgr.attr.ModifyEnergy(hit.Defender.ID(), hit.EnergyGain*hit.HitRatio)
+	}
+
 	mgr.event.HitEnd.Emit(event.HitEndEvent{
 		Attacker:         hit.Attacker.ID(),
 		Defender:         hit.Defender.ID(),
 		AttackType:       hit.AttackType,
 		DamageType:       hit.DamageType,
-		BaseDamage:       0, // TODO
-		BonusDamage:      0, // TODO
-		TotalDamage:      0, // TODO
-		ShieldDamage:     0, // TODO
-		HPDamage:         0, // TODO
+		HPDamage:         hpUpdate,
+		BaseDamage:       base,
+		BonusDamage:      bonus,
+		TotalDamage:      total,
+		ShieldDamage:     total,
 		HPRatioRemaining: mgr.attr.HPRatio(hit.Defender.ID()),
-		IsCrit:           false, // TODO
+		IsCrit:           crit,
 		UseSnapshot:      hit.UseSnapshot,
 	})
 }
 
+// BASE DAMAGE:
+// Base DMG = (MV + Extra MV) * Scaling Attribute + Extra DMG
+// k = scaling attribute
+// v = (MV + Extra MV)
+func (mgr *Manager) baseDamage(h *info.Hit) float64 {
+	var dmgMap info.DamageMap = h.BaseDamage
+	damage := 0.0
+	for k, v := range dmgMap {
+		switch k {
+		case model.DamageFormula_BY_ATK:
+			damage += v * h.Attacker.ATK()
+		case model.DamageFormula_BY_DEF:
+			damage += v * h.Attacker.DEF()
+		case model.DamageFormula_BY_MAX_HP:
+			damage += v * h.Attacker.MaxHP()
+		case model.DamageFormula_BY_BREAK_DAMAGE:
+			damage += v * breakBaseDamage[h.Attacker.Level()] // TODO: evaluate if this is the best way
+		}
+	}
+	return damage
+}
+
+func (mgr *Manager) crit(h *info.Hit) bool {
+	if h.AttackType == model.AttackType_DOT || h.AttackType == model.AttackType_ELEMENT_DAMAGE || h.AsPureDamage {
+		return false
+	}
+	return mgr.rdm.Float64() < h.Attacker.CritChance()
+}
+
+func (mgr *Manager) bonusDamage(h *info.Hit) float64 {
+	dmg := 1.0
+	// Checks if hit doesn't use ByPureDamage equation
+	// Adds bEffect if ByPureDamage, and dmg% if not.
+	if h.BaseDamage[model.DamageFormula_BY_BREAK_DAMAGE] < 0 {
+		dmg += h.Attacker.GetProperty(prop.AllDamagePercent)
+		dmg += h.Attacker.GetProperty(prop.DamagePercent(h.DamageType))
+		if h.AttackType == model.AttackType_DOT {
+			dmg += h.Attacker.GetProperty(prop.DOTDamagePercent)
+		}
+	} else {
+		dmg += h.Attacker.BreakEffect()
+	}
```

Comment:
* comment doesn't match logic here (no check for pure damage?)
* checking for if `BY_BREAK_DAMAGE` is not used should be
```go
if _, ok := h.BaseDamage[model.DamageFormula_BY_BREAK_DAMAGE]; !ok {
  ...
}
```